tests/unit_tests/dlopen_source/source_sampling.cpp

```diff
@@ -0,0 +1,23 @@
+#include <iostream>
+#include "openmc/random_lcg.h"
+#include "openmc/source.h"
+#include "openmc/particle.h"
+
+// you must have external C linkage here otherwise 
+// dlopen will not find the file
+extern "C" openmc::Particle::Bank sample_source() {
+    openmc::Particle::Bank particle;
+    // wgt
+    particle.particle = openmc::Particle::Type::neutron;
+    particle.wgt = 1.0;
+    // position
+    
+    particle.r.x = 0.;
+    particle.r.y = 0.;
+    particle.r.z = 0.;
```

Comment:
zeros will be the default for `particle.r`, so technically speaking you could remove these three lines. I understand if you want to keep them for clarity though